codex/sales/reservations.nim

```diff
@@ -0,0 +1,314 @@
+proc `==`*(x, y: Availability): bool =
+  x.id == y.id and
+  x.size == y.size and
+  x.duration == y.duration and
+  x.minPrice == y.minPrice
+
+proc toErr[E1: ref CatchableError, E2: AvailabilityError](
+  e1: E1,
+  _: type E2,
+  msg: string = "see inner exception"): ref E2 =
+
+  return newException(E2, msg, e1)
+
+proc writeValue*(
+  writer: var JsonWriter,
+  value: SlotId | AvailabilityId) {.upraises:[IOError].} =
+
+  mixin writeValue
+  writer.writeValue value.toArray
+
+proc readValue*[T: SlotId | AvailabilityId](
+  reader: var JsonReader,
+  value: var T) {.upraises: [SerializationError, IOError].} =
+
+  mixin readValue
+  value = T reader.readValue(T.distinctBase)
+
+func used*(availability: Availability): bool =
+  availability.slotId.isSome
+
+func key(id: AvailabilityId): ?!Key =
+  (ReservationsKey / id.toArray.toHex)
+
+func key*(availability: Availability): ?!Key =
+  return availability.id.key
+
+func available*(self: Reservations): uint =
+  return self.repo.quotaMaxBytes - self.repo.totalUsed
+
+func available*(self: Reservations, bytes: uint): bool =
+  return bytes < self.available()
+
+proc exists*(
+  self: Reservations,
+  id: AvailabilityId): Future[?!bool] {.async.} =
+
+  without key =? id.key, err:
+    return failure(err)
+
+  let exists = await self.repo.metaDs.contains(key)
+  return success(exists)
+
+proc get*(
+  self: Reservations,
+  id: AvailabilityId): Future[?!Availability] {.async.} =
+
+  if exists =? (await self.exists(id)) and not exists:
+    let err = newException(AvailabilityNotExistsError,
+      "Availability does not exist")
+    return failure(err)
+
+  without key =? id.key, err:
+    return failure(err)
+
+  without serialized =? await self.repo.metaDs.get(key), err:
+    return failure(err)
+
+  without availability =? Json.decode(serialized, Availability).catch, err:
+    return failure(err)
+
+  return success availability
+
+proc update(
+  self: Reservations,
+  availability: Availability,
+  slotId: ?SlotId): Future[?!void] {.async.} =
+
+  without var updated =? await self.get(availability.id), err:
+    return failure(err)
+
+  updated.slotId = slotId
+
+  without key =? availability.key, err:
+    return failure(err)
+
+  if err =? (await self.repo.metaDs.put(
+    key,
+    @(updated.toJson.toBytes))).errorOption:
+    return failure(err)
+
+  return success()
+
+proc reserve*(
+  self: Reservations,
+  availability: Availability): Future[?!void] {.async.} =
+
+  if exists =? (await self.exists(availability.id)) and exists:
+    let err = newException(AvailabilityAlreadyExistsError,
+      "Availability already exists")
+    return failure(err)
+
+  without key =? availability.key, err:
+    return failure(err)
+
+  if err =? (await self.repo.metaDs.put(
+    key,
+    @(availability.toJson.toBytes))).errorOption:
+    return failure(err)
+
+  # TODO: reconcile data sizes -- availability uses UInt256 and RepoStore
+  # uses uint, thus the need to truncate
```

Comment:
I thoug about changing this to UIn256 in the RepoStore, but 2^64 can represent 16 exabytes, which seems enough for the short term ;-)